File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        // three option
        
        
        function<int(TreeNode*)> DFS;
        int ans = INT_MIN;
        DFS = [&](TreeNode *root){
            if(root == NULL) return 0;
            
            int left = DFS(root->left);
            int right = DFS(root->right);
            
            int mx = max({root->val, root->val + left, root->val + right});
            ans = max({ans, mx, root->val + right + left});
            // cout << root->val << " " << mx << endl;
            return mx;
        };
        
        DFS(root);
        
        return ans;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:

        ans = float("-inf")

        def DFS(root: TreeNode | None):
            nonlocal ans

            if not root:
                return 0
    
            l = DFS(root.left)
            r = DFS(root.right) 
            c = root.val

            ans = max(ans, l + r + c, l + c, r + c, c)
            return max(l + c, r + c, c)
        
        DFS(root)

        return ans